File: forensics-btmgen/src/main/java/de/burger/forensics/plugin/engine/JavaRegexParser.java
package de.burger.forensics.plugin.engine;

import static de.burger.forensics.plugin.engine.JavaPrefilter.prefilterJava;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class JavaRegexParser implements JavaScanner {
    private static final Pattern PACKAGE_PATTERN =
            Pattern.compile("(?m)^\\s*package\\s+([a-zA-Z0-9_.]+)\\s*;");
    private static final Pattern CLASS_PATTERN = Pattern.compile(
            "(?m)^\\s*(?:@[\\w$.]+(?:\\([^)]*+\\))?\\s*)*(?:(?:\\b(?:public|protected|private|abstract|final|static|strictfp|sealed)\\b|non-sealed)\\s+)*class\\s+([A-Za-z0-9_]+)");
    private static final Pattern METHOD_PATTERN = Pattern.compile(
            "(?m)^\\s*(?:@[\\w$.]+(?:\\([^)]*+\\))?\\s*)*(?:\\b(?:public|protected|private|abstract|final|static|strictfp|synchronized|native|default)\\b\\s+)*(?:<[^>]*+>\\s*)?[\\w$<>\\[\\],.?\\s]+\\s+([a-zA-Z0-9_]+)\\s*\\(([^)]*+)\\)\\s*\\{");
    private static final Pattern IF_PATTERN = Pattern.compile("\\bif\\s*\\(([^)]*+)\\)");
    private static final Pattern SWITCH_PATTERN = Pattern.compile("\\bswitch\\s*\\(([^)]*+)\\)");
    private static final Pattern CASE_PATTERN = Pattern.compile("(?m)^[\\t ]*(case\\s+[^:]*+|default)\\s*:");
    private static final Pattern WHITESPACE_PATTERN = Pattern.compile("\\s+");

    @Override
    public List<String> scan(
            String text,
            String helperFqn,
            String packagePrefix,
            boolean includeEntryExit,
            int maxStringLength) {
        List<String> rules = new ArrayList<>();
        String sanitized = prefilterJava(text);
        LineIndex lineIndex = new LineIndex(text);

        Matcher pkgMatcher = PACKAGE_PATTERN.matcher(sanitized);
        String pkg = pkgMatcher.find() ? pkgMatcher.group(1) : "";
        if (packagePrefix != null
                && !packagePrefix.isBlank()
                && !pkg.isBlank()
                && !pkg.startsWith(packagePrefix)) {
            return Collections.emptyList();
        }

        Matcher classMatcher = CLASS_PATTERN.matcher(sanitized);
        while (classMatcher.find()) {
            String className = classMatcher.group(1);
            String fqcn = pkg.isBlank() ? className : pkg + "." + className;
            int openIndex = sanitized.indexOf('{', classMatcher.end());
            if (openIndex < 0) {
                continue;
            }
            int closeIndex = findMatchingBrace(sanitized, openIndex);
            String bodySanitized = sanitized.substring(openIndex + 1, closeIndex);
            Matcher methodMatcher = METHOD_PATTERN.matcher(bodySanitized);
            while (methodMatcher.find()) {
                String methodName = methodMatcher.group(1);
                int methodStartInClass = openIndex + 1 + methodMatcher.start();
                int methodOpen = sanitized.indexOf('{', methodStartInClass);
                if (methodOpen < 0) {
                    continue;
                }
                int methodClose = findMatchingBrace(sanitized, methodOpen);
                String methodBodySanitized = sanitized.substring(methodOpen + 1, methodClose);
                String methodBodyOriginal = text.substring(methodOpen + 1, methodClose);

                if (includeEntryExit) {
                    rules.add(buildEntryRule(helperFqn, fqcn, methodName));
                    rules.add(buildExitRule(helperFqn, fqcn, methodName));
                }

                Matcher ifMatcher = IF_PATTERN.matcher(methodBodySanitized);
                while (ifMatcher.find()) {
                    if (ifMatcher.group(1) == null) {
                        continue;
                    }
                    int offset = methodOpen + 1 + ifMatcher.start();
                    int line = lineIndex.lineAt(offset);
                    int condStart = ifMatcher.start(1);
                    int condEnd = ifMatcher.end(1);
                    String condRaw = methodBodyOriginal.substring(condStart, condEnd);
                    String cond = escape(condRaw, maxStringLength);
                    rules.add(
                            "RULE " + fqcn + "." + methodName + ":" + line + ":if-true\n"
                                    + "CLASS " + fqcn + "\n"
                                    + "METHOD " + methodName + "(..)\n"
                                    + "HELPER " + helperFqn + "\n"
                                    + "AT LINE " + line + "\n"
                                    + "IF (" + condRaw + ")\n"
                                    + "DO iff(\"" + fqcn + "\",\"" + methodName + "\"," + line + ",\""
                                    + cond + "\", true)\n"
                                    + "ENDRULE");
                    rules.add(
                            "RULE " + fqcn + "." + methodName + ":" + line + ":if-false\n"
                                    + "CLASS " + fqcn + "\n"
                                    + "METHOD " + methodName + "(..)\n"
                                    + "HELPER " + helperFqn + "\n"
                                    + "AT LINE " + line + "\n"
                                    + "IF (!(" + condRaw + "))\n"
                                    + "DO iff(\"" + fqcn + "\",\"" + methodName + "\"," + line + ",\""
                                    + cond + "\", false)\n"
                                    + "ENDRULE");
                }

                Matcher switchMatcher = SWITCH_PATTERN.matcher(methodBodySanitized);
                while (switchMatcher.find()) {
                    if (switchMatcher.group(1) == null) {
                        continue;
                    }
                    int offset = methodOpen + 1 + switchMatcher.start();
                    int line = lineIndex.lineAt(offset);
                    int selectorStart = switchMatcher.start(1);
                    int selectorEnd = switchMatcher.end(1);
                    String selectorRaw = methodBodyOriginal.substring(selectorStart, selectorEnd);
                    String sel = escape(selectorRaw, maxStringLength);
                    rules.add(
                            "RULE " + fqcn + "." + methodName + ":" + line + ":when\n"
                                    + "CLASS " + fqcn + "\n"
                                    + "METHOD " + methodName + "(..)\n"
                                    + "HELPER " + helperFqn + "\n"
                                    + "AT LINE " + line + "\n"
                                    + "DO sw(\"" + fqcn + "\",\"" + methodName + "\"," + line + ",\"" + sel
                                    + "\")\n"
                                    + "ENDRULE");
                }

                Matcher caseMatcher = CASE_PATTERN.matcher(methodBodySanitized);
                while (caseMatcher.find()) {
                    if (caseMatcher.group(1) == null) {
                        continue;
                    }
                    int labelStart = caseMatcher.start(1);
                    int labelEnd = caseMatcher.end(1);
                    String labelOriginal = methodBodyOriginal.substring(labelStart, labelEnd);
                    String label = WHITESPACE_PATTERN.matcher(labelOriginal).replaceAll(" ").trim();
                    int offset = methodOpen + 1 + caseMatcher.start();
                    int line = lineIndex.lineAt(offset);
                    String esc = escape(label, maxStringLength);
                    rules.add(
                            "RULE " + fqcn + "." + methodName + ":" + line + ":case\n"
                                    + "CLASS " + fqcn + "\n"
                                    + "METHOD " + methodName + "(..)\n"
                                    + "HELPER " + helperFqn + "\n"
                                    + "AT LINE " + line + "\n"
                                    + "DO kase(\"" + fqcn + "\",\"" + methodName + "\"," + line + ",\"" + esc
                                    + "\")\n"
                                    + "ENDRULE");
                }
            }
        }

        return rules;
    }

    private String escape(String value, int limit) {
        String truncated = value;
        if (limit > 0 && value.length() > limit) {
            truncated = value.substring(0, limit) + "…";
        }
        return truncated
                .replace("\\", "\\\\")
                .replace("\"", "\\\"")
                .replace("\n", "\\n")
                .replace("\r", "\\r")
                .replace("\t", "\\t");
    }

    private String buildEntryRule(String helper, String className, String methodName) {
        return "RULE enter@" + className + "." + methodName + "\n"
                + "CLASS " + className + "\n"
                + "METHOD " + methodName + "(..)\n"
                + "HELPER " + helper + "\n"
                + "AT ENTRY\n"
                + "DO enter(\"" + className + "\",\"" + methodName + "\", $LINE)\n"
                + "ENDRULE";
    }

    private String buildExitRule(String helper, String className, String methodName) {
        return "RULE exit@" + className + "." + methodName + "\n"
                + "CLASS " + className + "\n"
                + "METHOD " + methodName + "(..)\n"
                + "HELPER " + helper + "\n"
                + "AT EXIT\n"
                + "DO exit(\"" + className + "\",\"" + methodName + "\", $LINE)\n"
                + "ENDRULE";
    }

    private int findMatchingBrace(String text, int openIndex) {
        int depth = 0;
        for (int i = openIndex; i < text.length(); i++) {
            char ch = text.charAt(i);
            if (ch == '{') {
                depth++;
            } else if (ch == '}') {
                depth--;
                if (depth == 0) {
                    return i;
                }
            }
        }
        return text.length() - 1;
    }

    private static final class LineIndex {
        private final int[] starts;

        LineIndex(String text) {
            List<Integer> tmp = new ArrayList<>();
            tmp.add(0);
            for (int i = 0; i < text.length(); i++) {
                if (text.charAt(i) == '\n') {
                    tmp.add(i + 1);
                }
            }
            starts = new int[tmp.size()];
            for (int i = 0; i < tmp.size(); i++) {
                starts[i] = tmp.get(i);
            }
        }

        int lineAt(int offset) {
            if (starts.length == 0) {
                return 1;
            }
            int low = 0;
            int high = starts.length - 1;
            while (low <= high) {
                int mid = (low + high) >>> 1;
                int s = starts[mid];
                int next = mid + 1 < starts.length ? starts[mid + 1] : Integer.MAX_VALUE;
                if (offset < s) {
                    high = mid - 1;
                } else if (offset >= next) {
                    low = mid + 1;
                } else {
                    return mid + 1;
                }
            }
            return starts.length;
        }
    }
}

File: forensics-btmgen/src/main/kotlin/de/burger/forensics/plugin/engine/JavaRegexParser.kt
package de.burger.forensics.plugin.engine

import java.util.regex.Pattern

class JavaRegexParser : JavaScanner {
    override fun scan(
        text: String,
        helperFqn: String,
        packagePrefix: String?,
        includeEntryExit: Boolean,
        maxStringLength: Int
    ): List<String> {
        val rules = mutableListOf<String>()
        val sanitized = prefilterJava(text)
        val lineIndex = LineIndex(text)
        val pkgMatcher = PACKAGE_PATTERN.matcher(sanitized)
        val pkg = if (pkgMatcher.find()) pkgMatcher.group(1) else ""
        if (!packagePrefix.isNullOrBlank() && pkg.isNotBlank() && !pkg.startsWith(packagePrefix)) {
            return emptyList()
        }
        val classMatcher = CLASS_PATTERN.matcher(sanitized)
        while (classMatcher.find()) {
            val className = classMatcher.group(1)
            val fqcn = if (pkg.isBlank()) className else "$pkg.$className"
            val openIndex = sanitized.indexOf('{', classMatcher.end())
            if (openIndex < 0) continue
            val closeIndex = findMatchingBrace(sanitized, openIndex)
            val bodySanitized = sanitized.substring(openIndex + 1, closeIndex)
            val methodMatcher = METHOD_PATTERN.matcher(bodySanitized)
            while (methodMatcher.find()) {
                val methodName = methodMatcher.group(1)
                val methodStartInClass = openIndex + 1 + methodMatcher.start()
                val methodOpen = sanitized.indexOf('{', methodStartInClass)
                if (methodOpen < 0) continue
                val methodClose = findMatchingBrace(sanitized, methodOpen)
                val methodBodySanitized = sanitized.substring(methodOpen + 1, methodClose)
                val methodBodyOriginal = text.substring(methodOpen + 1, methodClose)
                if (includeEntryExit) {
                    rules += buildEntryRule(helperFqn, fqcn, methodName)
                    rules += buildExitRule(helperFqn, fqcn, methodName)
                }
                IF_REGEX.findAll(methodBodySanitized).forEach { match ->
                    val offset = methodOpen + 1 + match.range.first
                    val line = lineIndex.lineAt(offset)
                    val condRange = match.groups[1]?.range ?: return@forEach
                    val condRaw = methodBodyOriginal.substring(condRange)
                    val cond = escape(condRaw, maxStringLength)
                    rules += """
                        RULE ${fqcn}.${methodName}:${line}:if-true
                        CLASS $fqcn
                        METHOD ${methodName}(..)
                        HELPER $helperFqn
                        AT LINE $line
                        IF (${condRaw})
                        DO iff("$fqcn","$methodName",${line},"$cond", true)
                        ENDRULE
                    """.trimIndent()
                    rules += """
                        RULE ${fqcn}.${methodName}:${line}:if-false
                        CLASS $fqcn
                        METHOD ${methodName}(..)
                        HELPER $helperFqn
                        AT LINE $line
                        IF (!(${condRaw}))
                        DO iff("$fqcn","$methodName",${line},"$cond", false)
                        ENDRULE
                    """.trimIndent()
                }
                SWITCH_REGEX.findAll(methodBodySanitized).forEach { match ->
                    val offset = methodOpen + 1 + match.range.first
                    val line = lineIndex.lineAt(offset)
                    val selectRange = match.groups[1]?.range ?: return@forEach
                    val selectorRaw = methodBodyOriginal.substring(selectRange)
                    val sel = escape(selectorRaw, maxStringLength)
                    rules += """
                        RULE ${fqcn}.${methodName}:${line}:when
                        CLASS $fqcn
                        METHOD ${methodName}(..)
                        HELPER $helperFqn
                        AT LINE $line
                        DO sw("$fqcn","$methodName",${line},"${sel}")
                        ENDRULE
                    """.trimIndent()
                }
                CASE_REGEX.findAll(methodBodySanitized).forEach { match ->
                    val labelRange = match.groups[1]?.range ?: return@forEach
                    val labelOriginal = methodBodyOriginal.substring(labelRange)
                    val label = labelOriginal.replace(WHITESPACE_REGEX, " ").trim()
                    val offset = methodOpen + 1 + match.range.first
                    val line = lineIndex.lineAt(offset)
                    val esc = escape(label, maxStringLength)
                    rules += """
                        RULE ${fqcn}.${methodName}:${line}:case
                        CLASS $fqcn
                        METHOD ${methodName}(..)
                        HELPER $helperFqn
                        AT LINE $line
                        DO kase("$fqcn","$methodName",${line},"$esc")
                        ENDRULE
                    """.trimIndent()
                }
            }
        }
        return rules
    }

    private fun escape(value: String, limit: Int): String {
        val truncated = if (limit > 0 && value.length > limit) value.take(limit) + "…" else value
        return truncated
            .replace("\\", "\\\\")
            .replace("\"", "\\\"")
            .replace("\n", "\\n")
            .replace("\r", "\\r")
            .replace("\t", "\\t")
    }

    private fun buildEntryRule(helper: String, className: String, methodName: String): String =
        """
        RULE enter@${className}.${methodName}
        CLASS ${className}
        METHOD ${methodName}(..)
        HELPER ${helper}
        AT ENTRY
        DO enter("${className}","${methodName}", ${'$'}LINE)
        ENDRULE
        """.trimIndent()

    private fun buildExitRule(helper: String, className: String, methodName: String): String =
        """
        RULE exit@${className}.${methodName}
        CLASS $className
        METHOD ${methodName}(..)
        HELPER $helper
        AT EXIT
        DO exit("$className","$methodName", ${'$'}LINE)
        ENDRULE
        """.trimIndent()

    private fun findMatchingBrace(text: String, openIndex: Int): Int {
        var depth = 0
        for (i in openIndex until text.length) {
            when (text[i]) {
                '{' -> depth++
                '}' -> {
                    depth--
                    if (depth == 0) return i
                }
            }
        }
        return text.length - 1
    }

    private class LineIndex(text: String) {
        private val starts: IntArray
        init {
            val tmp = mutableListOf(0)
            text.forEachIndexed { idx, c -> if (c == '\n') tmp += idx + 1 }
            starts = tmp.toIntArray()
        }
        fun lineAt(offset: Int): Int {
            if (starts.isEmpty()) return 1
            var low = 0
            var high = starts.size - 1
            while (low <= high) {
                val mid = (low + high) ushr 1
                val s = starts[mid]
                val next = if (mid + 1 < starts.size) starts[mid + 1] else Int.MAX_VALUE
                when {
                    offset < s -> high = mid - 1
                    offset >= next -> low = mid + 1
                    else -> return mid + 1
                }
            }
            return starts.size
        }
    }

    private companion object {
        private val PACKAGE_PATTERN = Pattern.compile("(?m)^\\s*package\\s+([a-zA-Z0-9_.]+)\\s*;")
        private val CLASS_PATTERN = Pattern.compile(
            "(?m)^\\s*(?:@[\\w$.]+(?:\\([^)]*\\))?\\s*)*(?:(?:\\b(?:public|protected|private|abstract|final|static|strictfp|sealed)\\b|non-sealed)\\s+)*class\\s+([A-Za-z0-9_]+)"
        )
        private val METHOD_PATTERN = Pattern.compile(
            "(?m)^\\s*(?:@[\\w$.]+(?:\\([^)]*\\))?\\s*)*(?:(?:\\b(?:public|protected|private|abstract|final|static|strictfp|synchronized|native|default)\\b)\\s+)*(?:<[^>]+>\\s*)?[\\w$<>\\[\\],.?\\s]+\\s+([a-zA-Z0-9_]+)\\s*\\(([^)]*)\\)\\s*\\{"
        )
        private val IF_REGEX = Regex("\\bif\\s*\\((.*?)\\)")
        private val SWITCH_REGEX = Regex("\\bswitch\\s*\\((.*?)\\)")
        private val CASE_REGEX = Regex("(?m)^[\\t ]*(case\\s+[^:]+|default)\\s*:")
        private val WHITESPACE_REGEX = Regex("\\s+")
    }
}
